src/platform/EFR32/Efr32PsaOperationalKeystore.cpp

```diff
@@ -0,0 +1,446 @@
+    {
+        // Could not find a free spot in the map
+        return CHIP_ERROR_NO_MEMORY;
+    }
+
+    // Create new key on the old or found key ID
+    error = mPendingKeypair->Create(id, EFR32OpaqueKeyUsages::ECDSA_P256_SHA256);
+    if (error != CHIP_NO_ERROR)
+    {
+        // Try deleting and recreating this key since keys don't get wiped on factory erase yet
+        mPendingKeypair->Delete();
+        error = mPendingKeypair->Create(id, EFR32OpaqueKeyUsages::ECDSA_P256_SHA256);
+    }
+
+    if (error != CHIP_NO_ERROR)
+    {
+        ResetPendingKey();
+        return error;
+    }
+
+    // Set CSR and state
+    size_t csrLength = outCertificateSigningRequest.size();
+    error            = mPendingKeypair->NewCertificateSigningRequest(outCertificateSigningRequest.data(), csrLength);
+    if (error != CHIP_NO_ERROR)
+    {
+        ResetPendingKey();
+        return error;
+    }
+
+    outCertificateSigningRequest.reduce_size(csrLength);
+    mPendingFabricIndex = fabricIndex;
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR Efr32PsaOperationalKeystore::ActivateOpKeypairForFabric(FabricIndex fabricIndex,
+                                                                   const Crypto::P256PublicKey & nocPublicKey)
+{
+    VerifyOrReturnError(mIsInitialized, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(mPendingKeypair != nullptr, CHIP_ERROR_INVALID_FABRIC_INDEX);
+    VerifyOrReturnError(IsValidFabricIndex(fabricIndex) && (fabricIndex == mPendingFabricIndex), CHIP_ERROR_INVALID_FABRIC_INDEX);
+
+    // Validate public key being activated matches last generated pending keypair
+    VerifyOrReturnError(mPendingKeypair->Pubkey().Matches(nocPublicKey), CHIP_ERROR_INVALID_PUBLIC_KEY);
+
+    mIsPendingKeypairActive = true;
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR Efr32PsaOperationalKeystore::CommitOpKeypairForFabric(FabricIndex fabricIndex)
+{
+    VerifyOrReturnError(mIsInitialized, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(mPendingKeypair != nullptr, CHIP_ERROR_INVALID_FABRIC_INDEX);
+    VerifyOrReturnError(IsValidFabricIndex(fabricIndex) && (fabricIndex == mPendingFabricIndex), CHIP_ERROR_INVALID_FABRIC_INDEX);
+    VerifyOrReturnError(mIsPendingKeypairActive == true, CHIP_ERROR_INCORRECT_STATE);
+
+    // Add key association to key map
+    EFR32OpaqueKeyId id = mPendingKeypair->GetKeyId();
+
+    if (id == kEFR32OpaqueKeyIdUnknown || id == kEFR32OpaqueKeyIdVolatile)
+    {
+        ResetPendingKey();
+        return CHIP_ERROR_INTERNAL;
+    }
+
+    if (mKeyMapping[id - kEFR32OpaqueKeyIdPersistentMin] != kUndefinedFabricIndex)
+    {
+        ResetPendingKey();
+        return CHIP_ERROR_INTERNAL;
+    }
+
+    mKeyMapping[id - kEFR32OpaqueKeyIdPersistentMin] = fabricIndex;
+
+    // Persist key map
+    CHIP_ERROR error = EFR32Config::WriteConfigValueBin(EFR32Config::kConfigKey_OpKeyMap, mKeyMapping, sizeof(mKeyMapping));
```

Comment:
You should probably abstract storing/loading the map. Right now it is dependant on exact size of mKeyMapping, which may vary by build parameters.

Also:
```
#ifndef SL_MATTER_MAX_STORED_OP_KEYS
#define SL_MATTER_MAX_STORED_OP_KEYS (kMaxValidFabricIndex - kMinValidFabricIndex + 1)
#endif
```

^^^^ This causes memory/storage consumption much larger than the max number of supported fabrics. This is fine, but is a cost that could/should be described, since the key mapping will now be ~255 bytes in the class, which will be a static load.

Reply:
I've redone the keymap load/store action so that it is dynamically sized based on the larger of the already-stored keymap size or the current build's requested size (and tries to downsize when the requested size is reduced).

Please review and close the conversation if this satisfies your comment.